TASK: Extract services and convert http components to middlewares for Neos 7

mficzel opened the pull request:
Convert http components to middlewares to be compatible with Flow and Neos 7.

In addition instead of checking the flow session the request is validated against cookie and parameter allow and
ignore lists that can be defined via setting

```
    # requests have to fulfill certain conditions for beeing cached
    request:
      # !!! Only the http methods "GET" and "HEAD" are supported !!!

      # a request will only qualify for caching if it contains no cookieParams that
      # are not ignored.
      cookieParams:
        # ignored cookie params exclude cookies that are handled by the frontenecn
        # and are not relevant for the backend. A usecase would be gdpr consent cookies
        # if they are only used on the client side
        ignore: []

      # a request will only qualify for caching if it only contains queryParams that
      # are allowed or ignored. All other arguments will prevent caching.
      queryParams:
        # allowed params become part of the cache identifier, use this for
        # arguments that modify the reponse but still allow caching like pagination
        allow: []

        # ignored arguments are not part of the cache identifier but do not
        # prevent caching either. Use this for arguments that are meaningless for
        # the backend like utm_campaign
        ignore: []
```

The creation of the 304 state is removed as the standardsCompliance middleware handles this already.
Obviously the version requirement had to be raised.

Resolves #12, #7

mficzel committed:
TASK: Convert http components to middlewares

In addition instead of checking the flow session the request is validated against cookie and parameter allow and
ignore lists that can be defined via setting.

```
    # requests have to fulfill certain conditions for beeing cached
    request:
      # !!! Only the http methods "GET" and "HEAD" are supported !!!

      # a request will only qualify for caching if it contains no cookieParams that
      # are not ignored.
      cookieParams:
        # ignored cookie params exclude cookies that are handled by the frontenecn
        # and are not relevant for the backend. A usecase would be gdpr consent cookies
        # if they are only used on the client side
        ignore: []

      # a request will only qualify for caching if it only contains queryParams that
      # are allowed or ignored. All other arguments will prevent caching.
      queryParams:
        # allowed params become part of the cache identifier, use this for
        # arguments that modify the reponse but still allow caching like pagination
        allow: []

        # ignored arguments are not part of the cache identifier but do not
        # prevent caching either. Use this for arguments that are meaningless for
        # the backend like utm_campaign
        ignore: []
```

The creation of the 304 state is removed as the standardsCompliance middleware handles this already.
Obviously the version requirement had to be raised.

mficzel commented:
A thing we should discuss is where in the middleware chain this should be placed.

albe commented:
> where in the middleware chain this should be placed

`after trustedProxy`? Definitely as early as possible

mficzel commented:
@albe i would rather see this behind the standardCompliance as it uses the features of that for 304 states. Problem is that the middleware order behind trusted proxies is not really well defined yet with routing and standard compliance being both after trustedProxies

Would it make sense to put routing after standardCompliance?

albe commented:
> i would rather see this behind the standardCompliance

Right. But OTOH it would be nice to cache the standards compliant response. So maybe this still needs two middlewares, one for checking cache hits and one for caching responses?

> Would it make sense to put routing after standardCompliance?

Hm. I think that would not hurt. It just opens up more posibility for other middlewares to "un-compliant" the response

mficzel commented:
We cannot cache the standard compliant response as the age header of cached results changes and also the 304 state depends on the request and the given ETag header.

albe commented on Configuration/Settings.yaml:
```diff
@@ -6,20 +6,14 @@ Flowpack:
+      middlewares:
+        'fullPageCache_RequestCache':
+          middleware: 'Flowpack\FullPageCache\Middleware\RequestCacheMiddleware'
+          position: 'after session 20'
```
Note: The middleware that stores the response, needs to come "before" session, so that session MW can set the `Set-Cookie` header and the caching middleware can avoid caching that response

mficzel replied:
Actually that is why this currently is behind "session" since currently the flow session is asked for deciding about cachability. I already wondered whether it would be better to just look at the cookies of the request and put this before "session".

mficzel replied:
This is now adjusted and the request cache middleware does only look at the incoming cookies and arguments to decide wether a request is cacheable and thus this is now placed after trustedProxies now as @albe suggested.

albe commented on Configuration/Settings.yaml:
```diff
@@ -6,20 +6,38 @@ Flowpack:
+      # a request will only qualify for caching if it contains no cookieParams that
+      # are not ignored.
+      cookieParams:
+        # ignored cookie params exclude cookies that are handled by the frontenecn
```
```suggestion
        # ignored cookie params exclude cookies that are handled by the frontend
```

albe commented on README.md:
```diff
@@ -19,6 +19,34 @@ Flowpack:
+      # a request will only qualify for caching if it contains no cookieParams that
+      # are not ignored.
+      cookieParams:
+        # ignored cookie params exclude cookies that are handled by the frontenecn
```
```suggestion
        # ignored cookie params exclude cookies that are handled by the frontend
```

albe commented on composer.json:
```diff
@@ -4,8 +4,8 @@
     "require": {
-        "neos/neos": "~4.3.6 || ^5.0 || dev-master",
-        "guzzlehttp/psr7": "^1.4"
+        "neos/neos": "^6.0 || dev-master",
```
```suggestion
        "neos/neos": "^7.0 || dev-master",
```

albe reviewed (commented):
neos/neos version needs to be adjusted. Also two typos but otherwise looks good from what I can tell (didn't test it)

mficzel committed:
Apply suggestions from code review

Co-authored-by: Alexander Berl <a.berl@outlook.com>

mficzel committed:
BUGFIX: Use HeaderEnabled instead of HeaderLifetime to determine cachability

That way caching is supported even if no lifetime is set and the cache is valid ad infinitum

kitsunet commented:
Looks good to me, thanks for taking care <3

paavo commented:
Is a new release planned @kitsunet?

mficzel commented:
@paavo this is in 2.1 afaik

paavo commented:
> @paavo this is in 2.1 afaik

oh, my fault 🙈 
thanks @mficzel
